### ISSUE TITLE
`--traefik-loadbalancer-ip` flag in `epinio installation`

### ISSUE THREAD

**Issue by jimmykarily** [epinio/epinio#476]
## Why

Controlling epic #488 

Because before we can let cert-manager create any let's encrypt certs, the system-domain has to resolve to the load balancer IP address. Otherwise the self test will fail and the cert-manager won't be able to recover before certain time has passed.

We will split the traefik installation to a separate command (#477 ).
The `epinio install` command won't install traefik unless `--traefik-loadbalancer-ip` flag is used.  When it is set, it will be passed to the Epinio endpoint service in the LoadBalancerIP field. This setting is respected by most public cloud providers (but not by DigitalOcean).
If a public cloud provider doesn't respect this setting, the user will have to install the traefik using the separate command, connect the generated IP address with their system domain before they can proceed with `epinio install` (in which case they won't provide the `--traefik-loadbalancer-ip` flag).

## Flow

- The user buys a static IP address from the cloud provider (NOTE: This is not possible in DigitalOcean
but it's possible in most other public cloud providers).
- The user creates an A record to point their custom domain to that IP address
- The user runs `epinio install --traefik-system-domain mydomain.org --loadbalancer-ip 1.2.3.4`

The last command does a check to see if the system domain actually resolves to the given IP address.
If not, a warning is printed ([TBD]: exit then or continue?)

epinio install --with-traefik-loadbalancer-ip 1.2.3.4 --system-domain mydomain.org

**Comment by manno** [comment]
What about the simple dev use case with omg.howdoi.website?

`epinio install --with-traefik`?

**Comment by jimmykarily** [comment]
We can have the following flags:

```
--traefik-install # default false
--loadbalancer-ip # default not set
--system-domain # default 1.2.3.4.omg.howdoi.website
--production-certs # default false
```

Most combinations of these make sense except maybe for:

`--traefik-install false --loadbalancer-ip 1.2.3.4` because the ip is then not used anywhere (since we won't install traefik).

We should also make clear that the system-domain should already be pointing to the ip when both are set, otherwise we will hit the cert-manager issue. With the right validations and documentation we can get this right.

**Comment by jimmykarily** [comment]
Chris pointed out that it may be hard to recover if the user releases the IP address after installing Epinio. E.g.

- The user gets an Elastic IP address (AWS) to be used as `--loadbalancer-ip` in `epinio install`
- The user sets up system domain with this IP address
- The user installs Epinio
- The user releases that Elastic IP address
- Boom!

Epinio is no longer reachable and the loadbalancer ip address is hardcoded in our epinio service. The only way to fix this is to directly access the service and change the load balancer ip. We could implement a new epinio command for that but it certainly adds complexity.

Another approach (suggested both by Mario and Chris) is to use DNS challenge instead of the HTTP one (https://cert-manager.io/docs/configuration/acme/dns01/). This requires API access to the DNS provider. Many DNS providers allow remote configuration through some API but many don't.

One more approach could be to only use self-signed certificates when first installing Epinio and let the user upgrade to let's encrypt (or other signer) with an explicit Epinio command. E.g. `epinio certs-signer letsencrypt --update-existing`
This command would re-generate all certs using the letsencrypt signer (epinio server cert, app certs etc).
Or it could be per end point. E.g. something like this: `epinio app change-cert-signer letsencrypt myapp`

This would allow the user to setup another signer later on and be able to use it for new or existing applications.

**Comment by andreas-kupries** [comment]
@colstrom ^ :bell: Your opinion on "... only use self-signed certificates when first installing Epinio and let the user upgrade to LE ..." ?

**Comment by agracey** [comment]
I personally like the self-signed by default with an option to upgrade their LE setup. That seems the easier code to maintain longer term? 

I think I have a few questions:
- What's the brittleness of each option? (how easy would it be for a user to screw up either the setup or the maintenance?)
- What's the cost to change from one option to the other if we see that people don't like the option we picked?
- Are there any k8s features that might be coming in the future that would change any of this?

**Comment by agracey** [comment]
Another thing to think of is if a user wants to do SSL termination external to the cluster (which is what we did for the CAP sandbox). This would obviously not be something that Epinio can or should automate, I just want to make sure that whatever option we choose it doesn't exclude that use case.

**Comment by andreas-kupries** [comment]
I suspect that these general questions are better asked at #488, as they are about the entire set of tickets and not this specific one.

**Comment by colstrom** [comment]
> @colstrom ^ 🔔 Your opinion on "... only use self-signed certificates when first installing Epinio and let the user upgrade to LE ..." ?

I don't like self-signed certificates, and I can't think of a scenario where they're the right solution these days. If we're in a position where we need certificates that aren't valid beyond our borders, then we should be bootstrapping a private CA, and using that.

I'd recommend looking at [mkcert](https://github.com/FiloSottile/mkcert) as a dead-simple solution, or [cfssl](https://github.com/cloudflare/cfssl) for more complex scenarios. Also, [step-ca](https://github.com/smallstep/certificates) really should be considered, as it provides an ACME-compatible API, and thus can be dropped in wherever LetEncrypt would work, but also works in scenarios where LetsEncrypt is not viable due to network policies.

I have container images and scripts that I can contribute for any of these options, should the need arise.

**Comment by jimmykarily** [comment]
What makes a private CA better than self-signed certificates? The only difference I see is that the user will only have to trust the CA once (for all endpoints, epinio, apps etc) while with self-signed certificates the user has to accept each and every one. Is there something else I'm missing?

**Comment by kkaempf** [comment]
Enterprise customers usually want to use a private CA. :wink:

**Comment by jimmykarily** [comment]
> Enterprise customers usually want to use a private CA. wink

but that's not something we would bootstrap right? (or is it?). If the customers already have a CA then we should allow them to use it with Epinio. That's what #63 is all about. But if they don't have a CA, is there any reason why a private CA is better than self signed certificates, other than the one I mentioned above?

**Comment by colstrom** [comment]
> What makes a private CA better than self-signed certificates? The only difference I see is that the user will only have to trust the CA once (for all endpoints, epinio, apps etc) while with self-signed certificates the user has to accept each and every one. Is there something else I'm missing?

Consistency.

With self-signed certs, there are at least two paths that have to be considered *in each place a secure connection is involved*, with one path validating (the default, in nearly all TLS libraries), and the other path ignoring validation.

With a private CA, there is exactly one case that must be handled: trusting the CA. This is done once *for each thing that involves secure connections* of any sort, and nearly always can be done in at the *system* level, rather than the *application* level, and as a result, this can be solved in a reusable manner. There are exceptions here, like if you're building custom trust stores in memory or something, but how often those *that* come up, right? ;p

More importantly, the scenario described for private CAs has to be handled *anyways*. Enterprise deployments often involve private CAs. We need to support that. This means we need to be testing this scenario anyways, which kind of implies that we need a way to bootstrap a private CA, right? Effectively, this should simply mean letting customers do things the way we test them. Unless we don't test private CAs, in which case I'd argue that it's even more compelling that we do this, if only to test a scenario we expect to be used.

TL;DR: It's 2021. Private CAs are trivial to set up because modern tools are less horrible for this sort of thing. Self-signed certs are a relic of a distant age best forgotten.

**Comment by rohitsakala** [comment]
So, I tried specifying `loadBalancerIP` field in spec of `traefik`service as POC and it worked. I could install èpinio` in one command. 

Note: Use regional static IP address (region should be that one where your kubernetes cluster resides).
